openhtf/__init__.py

```diff
@@ -105,31 +118,79 @@ def code_info(self):
   def metadata(self):
     return self._test_info.metadata
 
-  def Execute(self, loop=None,
-              test_start=triggers.AutoStart, test_stop=triggers.AutoStop,
-              http_port=http_api.DEFAULT_HTTP_PORT):
+  def Configure(self, **kwargs):
+    """Update test-wide configuration options.
+
+    Valid kwargs:
+      http_port: Port on which to run the http_api, or None to disable.
+      output_callbacks: List of output callbacks to run, typically it's better
+          to use AddOutputCallbacks(), but you can pass [] here to reset them.
+    """
+    for key, value in kwargs.iteritems():
+      setattr(self._test_options, key, value)
+
+  def Stop(self, *_):
+    _LOG.error('Received SIGINT.')
+    with self._lock:
+      if self._executor:
+        # TestState str()'s nicely to a descriptive string, so let's log that
+        # just for good measure.
+        _LOG.error('Stopping Test due to SIGINT: %s', self._executor.GetState())
+        self._executor.Stop()
+      else:
+        # Once we're stopped for SIGINT, don't try to start anything else.
+        _LOG.error('No Test running, ignoring SIGINT.')
+        self._stopped = True
+
+  def Execute(self, test_start=lambda: 'UNKNOWN_DUT_ID', loop=None):
     """Starts the framework and executes the given test.
+
     Args:
-      test_start: Trigger for starting the test, defaults to AutoStart with a
+      test_start: Trigger for starting the test, defaults to a lambda with a
           dummy serial number.
-      test_stop: Trigger for when the test is over, defaults to AutoStop to
-          immediately stop after the phase.
-      output_callbacks: List of callbacks to be called with the results
-          output from this test.
+      loop: DEPRECATED
     """
-    SetupFramework()
-
+    # TODO(madsci): Remove this after a transitionary period.
     if loop is not None:
-      self.loop = loop
-
-    _LOG.info('Executing test: %s', self.code_info.name)
-    executor = exe.TestExecutor(self, test_start, test_stop)
-    http_server = http_api.Server(executor, http_port)
-    StopOnSigInt([http_server.Stop, executor.Stop])
-    http_server.Start()
-    executor.Start()
-    executor.Wait()
-    http_server.Stop()
+      raise ValueError(
+          'DEPRECATED. Looping is no longer natively supported by OpenHTF, '
+          'use a while True: loop around Test.Execute() instead.')
+
+    # These internally ensure they are safe to call multiple times with no weird
+    # side effects.
+    logs.SetupLogger()
+    CreateArgParser().parse_args()
```

Comment:
there are side-effects here though, if you tried to override a value from a flag, the flag will be reparsed and set again. I had put this in `SetupFramework()` so the user could call that early, then do what they wanted with how things were setup, then call `Execute()` as necessary